opm/common/OpmLog/MessageLimiter.hpp

```diff
@@ -46,62 +54,108 @@ namespace Opm
         {
-            if (tag.empty() || message_limit_ == NoLimit) {
-                return Response::PrintMessage;
+            Response res = Response::PrintMessage;
+            // Deal with tag limits.
+            if (tag.empty() || tag_limit_ == NoLimit) {
+                category_counts_[messageMask]++;
+                res = Response::PrintMessage;
             } else {
                 // See if tag already encountered.
                 auto it = tag_counts_.find(tag);
                 if (it != tag_counts_.end()) {
                     // Already encountered this tag. Increment its count.
                     const int count = ++it->second;
-                    return countBasedResponse(count);
+                    res = countBasedResponse(count, messageMask);
                 } else {
                     // First encounter of this tag. Insert 1.
                     tag_counts_.insert({tag, 1});
-                    return countBasedResponse(1);
+                    res = countBasedResponse(1, messageMask);
                 }
             }
+            // If tag count reached the limit, ignore all the same tagged messages.
+            if (res == Response::JustOverTagLimit || res == Response::OverTagLimit) {
+                return res;
+            } else {
+                if (category_limits_[messageMask] == NoLimit) {
+                    category_counts_[messageMask]++;
+                    res = Response::PrintMessage;
+                } else {
+                    res = countBasedResponse(messageMask);
+                }
+            }
+            return res;
         }
 
     private:
-        Response countBasedResponse(const int count)
+        Response countBasedResponse(const int count, const int64_t messageMask)
         {
-            if (count <= message_limit_) {
+            if (count <= tag_limit_) {
+                category_counts_[messageMask]++;
                 return Response::PrintMessage;
-            } else if (count == message_limit_ + 1) {
-                return Response::JustOverLimit;
+            } else if (count == tag_limit_ + 1) {
+                category_counts_[messageMask]++;
+                return Response::JustOverTagLimit;
             } else {
-                return Response::OverLimit;
+                return Response::OverTagLimit;
             }
         }
 
 
-        int message_limit_;
+        Response countBasedResponse(const int64_t messageMask)
+        {
+            if (category_counts_[messageMask] < category_limits_[messageMask]) {
+                category_counts_[messageMask]++;
+                return Response::PrintMessage;
+            } else if (category_counts_[messageMask] == category_limits_[messageMask] + 1) {
+                category_counts_[messageMask]++;
+                return Response::JustOverCategoryLimit;
+            } else {
+                return Response::OverCategoryLimit;
+            }
+        }
```

Comment:
I think the division of responsibility between handleMessageLimits() and the countBasedResponse() overloads is unclear. In the original code, countBasedResponse() did not change any state of the class (it could have been a const method, it was just an oversight that it was not). Here however, the category_counts_ get incremented. All in all, I have trouble following the logic about where to increase this counter. Please rewrite these parts so that the counter is only incremented in a single place.